mysql/mysqldump.go

```diff
@@ -0,0 +1,578 @@
+	}
+	l = append(l, table.Name.String())
+	return strings.Join(l, "."), nil
+}
+
+func processColumn(conv *internal.Conv, tableName string, col *ast.ColumnDef) (string, schema.Column, bool, error) {
+	if col.Name == nil {
+		return "", schema.Column{}, false, fmt.Errorf("column name is nil")
+	}
+	name := col.Name.OrigColName()
+	if col.Tp == nil {
+		return "", schema.Column{}, false, fmt.Errorf("can't get column type for %s: %w", name, fmt.Errorf("found nil *ast.ColumnDef.Tp"))
+	}
+	tid, mods := getTypeModsAndID(col.Tp.String())
+	ty := schema.Type{
+		Name:        tid,
+		Mods:        mods,
+		ArrayBounds: getArrayBounds(col.Tp.String(), col.Tp.Elems)}
+	column := schema.Column{Name: name, Type: ty}
+	return name, column, updateColsByOption(conv, tableName, col, &column), nil
+}
+
+// updateColsByOption is specifially for ColDef constraints.
+// ColumnOption type is used for parsing column constraint info from MySQL
+func updateColsByOption(conv *internal.Conv, tableName string, col *ast.ColumnDef, column *schema.Column) bool {
+	isPk := false
+	for _, elem := range col.Options {
+		switch op := elem.Tp; op {
+		case ast.ColumnOptionPrimaryKey:
+			column.NotNull = true
+			// If primary key is defined in a column then `isPk` will be true
+			// and this column will be added in colDef as primary keys
+			isPk = true
+		case ast.ColumnOptionNotNull:
+			column.NotNull = true
+		case ast.ColumnOptionDefaultValue:
+			column.Ignored.Default = true
+		case ast.ColumnOptionUniqKey:
+			column.Unique = true
+		case ast.ColumnOptionCheck:
+			column.Ignored.Check = true
+		case ast.ColumnOptionReference:
+			column.Ignored.ForeignKey = true
+		default:
+			conv.Unexpected(fmt.Sprintf("Not supported type option of const value `%v` found in table : %s , column : %s. Mapping of columnOptionType can be found here: 'https://godoc.org/github.com/pingcap/parser/ast#ColumnOptionType'", op, tableName, column.Name))
+		}
+	}
+	return isPk
+}
+
+// getTypeModsAndID returns ID and mods of column datatype
+func getTypeModsAndID(columnType string) (string, []int64) {
+	// There are no methods in pincap parser to retirieve ID and mods
+	// We will process columnType eg:'varchar(40)' and split ID from the string.
+	// We retrieve mods using regex expression and convert it to INT64
+	id := columnType
+	var mods []int64
+	if strings.Contains(columnType, "(") {
+		id = strings.Split(columnType, "(")[0]
+		// For 'set' and 'enum' datatypes, values provided are not maxLength
+		if id == "set" || id == "enum" {
+			return id, nil
+		}
+		values := valuesRegexp.FindString(columnType)
+		strMods := strings.Split(values[1:len(values)-1], ",")
+		for _, i := range strMods {
+			j, err := strconv.ParseInt(i, 10, 64)
+			if err != nil {
+				panic(err)
+			}
+			mods = append(mods, j)
+		}
+	}
+	// 'BINARY' keyword will be added to all blob datatypes by parser.
+	// Eg: mediumblob BINARY. It needs to be ignored to retrieve ID
+	if strings.Contains(id, " ") {
```

Comment:
Is this too generic?
This will split any type id that contains " ", not just something with " BINARY".

Is it possible for a type id to have a " " in it, except for this case?
I was thinking of user defined types (combined the fact that mysql allows you to put anything in an id if you quote it e.g. `my crazy; col` is a valid column name).
But I see that mysql doesn't support user defined types, so maybe not an issue.

Maybe be more targeted here and use strings.TrimSuffix?

Reply:
Correct, mysql doesn't support user defined datatypes. We have tested using all the datatypes under mysql and `blob` is an exception in which there is a " " followed by `BINARY` keyword which is added by pingcap parser.
Got it. I will make it more targeted by using `id = string.TrimSuffix(columnType," BINARY")`

Reply:
https://github.com/cloudspannerecosystem/harbourbridge/blob/99db369b2b826fc24ffbb760f3b3700cd0e04b0f/mysql/mysqldump.go#L377-L381
Done.